cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetTypeService.java

```diff
@@ -0,0 +1,417 @@
+    DatasetModuleMeta moduleMeta = typeManager.getModule(moduleId);
+    if (moduleMeta == null) {
+      throw new DatasetModuleNotFoundException(moduleId);
+    }
+    Principal principal = authenticationContext.getPrincipal();
+    final Predicate<EntityId> filter = authorizationEnforcer.createFilter(principal);
+    if (!Principal.SYSTEM.equals(principal) && !filter.apply(datasetModuleId)) {
+      throw new UnauthorizedException(principal, datasetModuleId);
+    }
+    return moduleMeta;
+  }
+
+  /**
+   * Adds a new {@link DatasetModule}.
+   *
+   * @param datasetModuleId the {@link DatasetModuleId} for the module to be added
+   * @param className the module class name specified in the HTTP header
+   * @param forceUpdate if true, an update will be allowed even if there are conflicts with other modules, or if
+   *                     removal of a type would break other modules' dependencies
+   * @return a {@link BodyConsumer} to upload the module jar in chunks
+   * @throws NotFoundException if the namespace in which the module is being added is not found
+   * @throws IOException if there are issues while performing I/O like creating temporary directories, moving/unpacking
+   *                      module jar files
+   * @throws DatasetModuleConflictException if #forceUpdate is {@code false}, and there are conflicts with other modules
+   */
+  BodyConsumer addModule(final DatasetModuleId datasetModuleId, final String className,
+                         final boolean forceUpdate) throws Exception {
+    NamespaceId namespaceId = datasetModuleId.getParent();
+    final Principal principal = authenticationContext.getPrincipal();
+    // enforce that the principal has WRITE access on the namespace
+    authorizationEnforcer.enforce(namespaceId, principal, Action.WRITE);
+    if (NamespaceId.SYSTEM.equals(namespaceId)) {
+      throw new UnauthorizedException(String.format("Cannot add module '%s' to '%s' namespace.",
+                                                    datasetModuleId.getModule(), datasetModuleId.getNamespace()));
+    }
+    ensureNamespaceExists(namespaceId);
+
+    // It is now determined that a new dataset module will be deployed. First grant privileges, then deploy the module.
+    // If creation fails, revoke the granted privileges. This ensures that just like delete, there may be orphaned
+    // privileges in rare scenarios, but there can never be orphaned datasets.
+    // If the module previously existed and was deleted, but revoking privileges somehow failed, there may be orphaned
+    // privileges for the module. Revoke them first, so no users unintentionally get privileges on the dataset.
+    revokeAllPrivilegesOnModule(datasetModuleId);
+    grantAllPrivilegesOnModule(datasetModuleId, principal);
+    try {
+      return createModuleConsumer(datasetModuleId, className, forceUpdate, principal);
+    } catch (Exception e) {
+      revokeAllPrivilegesOnModule(datasetModuleId);
```

Comment:
yup, if anyone catches this exception, they will log. If it propagates to the router, ``HTTPExceptionHandler`` will include only the exception message in the response body with a 500 response code.